### ISSUE TITLE
Calling `setProps` on wrapperresults in `componentDidUpdate` being called twice

### ISSUE THREAD

**Issue by jpmoral** [airbnb/enzyme#1862]
**Describe the bug**
Calling `setProps` results in `componentDidUpdate` being called twice (instead of once).

**To Reproduce**
Run the test:

```
import React from 'react';


export default class Dummy extends React.Component {
    componentWillReceiveProps(nextProps) {
        this.setState({someState: nextProps.myProp});
    }

    componentDidUpdate() {
        this.foo();
    }

    foo() {}

    render() {
        return (
            <div>
                <div>{`myProp: ${this.props.myProp}`}</div>
            </div>
        );
    }
}
```

```
import React from 'react';
import {shallow} from 'enzyme';

import Dummy from 'questions/components/dummy.js';


describe('<Dummy/>', () => {
    it('works', () => {
        const props = {};
        const wrapper = shallow(<Dummy {...props} />);
        wrapper.instance().foo = jest.fn();
        wrapper.setProps({myProp: 'Prop Value'});
        expect(wrapper.instance().foo).toHaveBeenCalledTimes(1);
    });
});
```

**Expected behavior**
`componentDidUpdate` should only be called once.

**Desktop (please complete the following information):**
 - OS: macOS Mojave
 - Browser n/a
 - Version 3.7.0

**Smartphone (please complete the following information):**
n/a

**Comment by koba04** [comment]
@ljharb 
Yeah, React doesn't support batched updates on ShallowRendering from v16.

https://github.com/facebook/react/issues/10355

So currently, it's a no-op.

https://github.com/airbnb/enzyme/blob/master/packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js#L327-L330

If we support batching updates on `shallow` to fix this, we have to implement ourselves.

**Comment by ljharb** [comment]
Although that's a lot more work, I think it's probably worth it so that things are consistent between React and enzyme.

**Comment by jpmoral** [comment]
@ljharb thanks, `mount` works. I've edited the title and body to explicitly mention `shallow`.

**Comment by peanutenthusiast** [comment]
I'll go ahead and look into this issue if a fix hasn't already been applied.

**Comment by peanutenthusiast** [comment]
@ljharb Added the above test case, but in terms of the fix, has there been much discussion on adding support for batched updates with the `ReactSixteenThreeAdapter`?  

I've been looking at @koba04's implementation of `unstable_batchedUpdates`

https://github.com/koba04/react/commit/ca78c92635525ad2dfd8012e7e5527a5876ce130,

but would you have any suggestions for moving forward?

**Comment by ljharb** [comment]
@peanutenthusiast not much, but I'd love to add it (ideally separate from bug fixes)

**Comment by peanutenthusiast** [comment]
Was out of country for a month or so, but I've come up with a fix for the above test case -- @ljharb are there any other test cases that should be covered? I can send a PR tonight.

**Comment by jpmoral** [comment]
Awesome, thanks everyone!
